frontend/components/instruments/StepSequencer.js

```diff
@@ -0,0 +1,229 @@
+import PropTypes from "prop-types";
+import SamplePlayer from "./SamplePlayer";
+import STYLES from "./StepSequencer.module.scss";
+
+const Step = ({sample, audioContext, colIsPlaying, vol}) => {
+    const [isActive, setIsActive] = useState(false);
+
+    const handleClick = () => {
+        setIsActive(!isActive);
+    };
+
+    // Pick button style depending on if sample is loaded
+    const btnStyle =
+        isActive
+            ? STYLES.activeStep
+            : STYLES.inactiveStep;
+
+    return (<>
+        <button
+            className={btnStyle}
+            onClick={handleClick}
+        />
+        <SamplePlayer
+            sample={sample}
+            shouldPlay={colIsPlaying && isActive}
+            loop={false}
+            volume={vol}
+            audioContext={audioContext}
+        />
+    </>);
+};
+Step.propTypes = {
+    sample: PropTypes.string,
+    colIsPlaying: PropTypes.bool,
+    audioContext: PropTypes.object,
+};
+
+const Metronome = ({sample, audioContext, colIsPlaying}) => {
+    // Pick button style depending on if sample is loaded
+    const btnStyle =
+        colIsPlaying
+            ? STYLES.isPlaying
+            : STYLES.notPlaying;
+
+    return (<>
+        <button
+            className={btnStyle}
+        />
+        <SamplePlayer
+            sample={sample}
+            shouldPlay={colIsPlaying}
+            loop={false}
+            volume={100}
+            audioContext={audioContext}
+        />
+    </>);
+};
+Metronome.propTypes = {
+    sample: PropTypes.string,
+    colIsPlaying: PropTypes.bool,
+    audioContext: PropTypes.object,
+};
+
+const Row = ({sample, audioContext, rowIndex, numSteps, playbackPosition, isPlaying}) => {
+    const [volume, setVolume] = useState(100);
+
+    const handleUpdateVolume = (event) => {
+        event.preventDefault();
+        if (!isPlaying){ // can only move slider when paused
+            setVolume(parseInt(event.target.value));
+        }
+    };
+
+    return (<>
+        <div>
+            {Array(numSteps).fill(null).map((_, colIndex) => (
+                <Step
+                    key={colIndex}
+                    sample={sample}
+                    audioContext={audioContext}
+                    colIsPlaying={isPlaying && colIndex === playbackPosition}
+                    vol = {volume}
+                />
+            ))}
+            Volume {rowIndex}
+            <input className="mx-2" type="range" min="0" max="100"
+                   step="1" id="low" value={volume}
+                   onChange={handleUpdateVolume}/>
+        </div>
+    </>);
+};
+Row.propTypes = {
+    sample: PropTypes.string,
+    audioContext: PropTypes.object,
+    rowIndex: PropTypes.number,
+    numSteps: PropTypes.number,
+    playbackPosition: PropTypes.number,
+    isPlaying: PropTypes.bool,
+};
+
+const StepSequencer = ({samples}) => {
+    // - stretch goal: metronome ticks even without playback--DONE
```

Comment:
Probably should cut these TODOs and open a follow-up issue for further work at this point.